File: src/main/scala/mffs/gui/GuiFortronCapacitor.java
package mffs.gui;

import cpw.mods.fml.common.network.PacketDispatcher;
import mffs.ModularForceFieldSystem;
import mffs.base.GuiMFFS;
import mffs.base.TileMFFS.TilePacketType;
import mffs.container.ContainerFortronCapacitor;
import mffs.gui.button.GuiButtonPressTransferMode;
import mffs.tile.TileFortronCapacitor;
import net.minecraft.client.gui.GuiButton;
import net.minecraft.entity.player.EntityPlayer;
import org.lwjgl.opengl.GL11;
import universalelectricity.api.energy.UnitDisplay;
import universalelectricity.api.energy.UnitDisplay.Unit;
import universalelectricity.api.energy.UnitDisplay.UnitPrefix;
import universalelectricity.api.vector.Vector2;

public class GuiFortronCapacitor extends GuiMFFS
{
	private TileFortronCapacitor tileEntity;

	public GuiFortronCapacitor(EntityPlayer player, TileFortronCapacitor tileentity)
	{
		super(new ContainerFortronCapacitor(player, tileentity), tileentity);
		this.tileEntity = tileentity;
	}

	@Override
	public void initGui()
	{
		this.textFieldPos = new Vector2(50, 76);
		super.initGui();
		this.buttonList.add(new GuiButtonPressTransferMode(1, this.width / 2 + 15, this.height / 2 - 37, this, this.tileEntity));

	}

	@Override
	protected void drawGuiContainerForegroundLayer(int x, int y)
	{
		this.fontRenderer.drawString(this.tileEntity.getInvName(), this.xSize / 2 - this.fontRenderer.getStringWidth(this.tileEntity.getInvName()) / 2, 6, 4210752);

		GL11.glPushMatrix();
		GL11.glRotatef(-90, 0, 0, 1);
		this.drawTextWithTooltip("upgrade", -95, 140, x, y);
		GL11.glPopMatrix();

		this.drawTextWithTooltip("linkedDevice", "%1: " + this.tileEntity.getLinkedDevices().size(), 8, 28, x, y);
		this.drawTextWithTooltip("transmissionRate", "%1: " + UnitDisplay.getDisplayShort(this.tileEntity.getTransmissionRate() * 20, Unit.LITER) + "/s", 8, 40, x, y);
		this.drawTextWithTooltip("range", "%1: " + this.tileEntity.getTransmissionRange(), 8, 52, x, y);
		this.drawTextWithTooltip("frequency", "%1:", 8, 63, x, y);
		this.textFieldFrequency.drawTextBox();
		this.drawTextWithTooltip("fortron", "%1:", 8, 95, x, y);
		this.fontRenderer.drawString(UnitDisplay.getDisplayShort(this.tileEntity.getFortronEnergy(), Unit.LITER) + "/" + UnitDisplay.getDisplay(this.tileEntity.getFortronCapacity(), Unit.LITER, UnitPrefix.MILLI), 8, 105, 4210752);

		if (tileEntity.getFortronCost() > 0)
		{
			fontRenderer.drawString("\u00a74-" + UnitDisplay.getDisplayShort(tileEntity.getFortronCost() * 20, Unit.LITER) + "/s", 118, 116, 4210752);
		}

		super.drawGuiContainerForegroundLayer(x, y);
	}

	@Override
	protected void drawGuiContainerBackgroundLayer(float f, int x, int y)
	{
		super.drawGuiContainerBackgroundLayer(f, x, y);

		/**
		 * Upgrade Slots
		 */
		this.drawSlot(153, 46);
		this.drawSlot(153, 66);
		this.drawSlot(153, 86);

		/**
		 * Frequency Card Slots
		 */
		this.drawSlot(8, 73);
		this.drawSlot(26, 73);

		this.drawForce(8, 115, Math.min((float) this.tileEntity.getFortronEnergy() / (float) this.tileEntity.getFortronCapacity(), 1));
	}

	@Override
	protected void actionPerformed(GuiButton guibutton)
	{
		super.actionPerformed(guibutton);

		if (guibutton.id == 1)
		{
			PacketDispatcher.sendPacketToServer(ModularForceFieldSystem.PACKET_TILE.getPacket(this.tileEntity, TilePacketType.TOGGLE_MODE.ordinal()));
		}
	}
}

File: src/main/scala/mffs/production/GuiFortronCapacitor.scala
package mffs.production

import mffs.ModularForceFieldSystem
import mffs.base.{GuiMFFS, TilePacketType}
import mffs.render.button.GuiTransferModeButton
import net.minecraft.client.gui.GuiButton
import net.minecraft.entity.player.EntityPlayer
import org.lwjgl.opengl.GL11
import resonant.lib.gui.GuiContainerBase.SlotType
import resonant.lib.network.PacketTile
import resonant.lib.render.EnumColor
import resonant.lib.wrapper.WrapList._
import universalelectricity.api.UnitDisplay

class GuiFortronCapacitor(player: EntityPlayer, tile: TileFortronCapacitor) extends GuiMFFS(new ContainerFortronCapacitor(player, tile), tile)
{
  override def initGui()
  {
    super.initGui()
    this.buttonList.add(new GuiTransferModeButton(1, this.width / 2 - 30, this.height / 2 - 30, this, this.tile))
  }

  protected override def drawGuiContainerForegroundLayer(x: Int, y: Int)
  {
    fontRendererObj.drawString(tile.getInventoryName, this.xSize / 2 - fontRendererObj.getStringWidth(tile.getInventoryName) / 2, 6, 4210752)
    GL11.glPushMatrix
    GL11.glRotatef(-90, 0, 0, 1)
    drawTextWithTooltip("upgrade", -95, 140, x, y)
    GL11.glPopMatrix
    drawTextWithTooltip("linkedDevice", "%1: " + tile.getDeviceCount, 8, 20, x, y)
    drawTextWithTooltip("transmissionRate", "%1: " + new UnitDisplay(UnitDisplay.Unit.LITER, tile.getTransmissionRate * 20).symbol() + "/s", 8, 32, x, y)
    drawTextWithTooltip("range", "%1: " + tile.getTransmissionRange, 8, 44, x, y)
    drawTextWithTooltip("input", EnumColor.DARK_GREEN + "%1", 12, 62, x, y)
    drawTextWithTooltip("output", EnumColor.RED + "%1", 92, 62, x, y)
    drawFortronText(x, y)
    super.drawGuiContainerForegroundLayer(x, y)
  }

  protected override def drawGuiContainerBackgroundLayer(f: Float, x: Int, y: Int)
  {
    super.drawGuiContainerBackgroundLayer(f, x, y)

    //Upgrade slots
    (0 to 2) foreach (y => drawSlot(153, 46 + y * 18))

    //Input slots
    for (x <- 0 to 1; y <- 0 to 1)
      drawSlot(8 + x * 18, 73 + y * 18)

    //Output slots
    for (x <- 0 to 1; y <- 0 to 1)
      drawSlot(90 + x * 18, 73 + y * 18)

    drawFrequencyGui()
  }

  protected override def actionPerformed(guiButton: GuiButton)
  {
    super.actionPerformed(guiButton)

    if (guiButton.id == 1)
    {
      ModularForceFieldSystem.packetHandler.sendToAll(new PacketTile(this.tile, TilePacketType.TOGGLE_MODE.id: Integer))
    }
  }
}